File: problems/11_graphs/0210_course_schedule_ii/solution.py
"""
Problem: 0210. Course Schedule II
Difficulty: Medium
URL: https://leetcode.com/problems/course-schedule-ii/

解法：拓撲排序 (Kahn)。
"""

from __future__ import annotations

from typing import List

from collections import deque

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        pass


def _is_valid_order(order: List[int], n: int, prereqs: List[List[int]]) -> bool:
    if len(order) != n:
        return False
    if set(order) != set(range(n)):
        return False
    pos = {v: i for i, v in enumerate(order)}
    for a, b in prereqs:
        if pos[b] > pos[a]:
            return False
    return True

if __name__ == "__main__":
    solution = Solution()

    # Test 1
    # result = solution.solve(...)
    # assert result == expected

    print("All tests passed!")

File: problems/11_graphs/0210_course_schedule_ii/solution.dart
// Problem: 0210. Course Schedule II
// Difficulty: Medium

import 'dart:collection';

List<int> findOrder(int numCourses, List<List<int>> prerequisites) {
  // TODO: implement
  return [];
}

bool _validOrder(List<int> order, int n, List<List<int>> prereqs) {
  if (order.length != n) return false;
  final seen = <int>{};
  for (final v in order) seen.add(v);
  if (seen.length != n) return false;
  final pos = <int, int>{};
  for (var i = 0; i < order.length; i++) pos[order[i]] = i;
  for (final p in prereqs) {
    if ((pos[p[1]] ?? -1) > (pos[p[0]] ?? -1)) return false;
  }
  return true;
}

void main() {
  // Test 1
  // final result = solve(...);
  print('All tests passed!');
}
